pkg/cli/clusteroperator/copytonode/copy_to_node.go

```diff
@@ -0,0 +1,166 @@
+package copytonode
+
+import (
+	"context"
+	_ "embed"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/openshift/oc/pkg/cli/clusteroperator/pernodepod"
+
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+)
+
+var (
+	//go:embed copy-pod-template.yaml
+	podYaml []byte
+	pod     = resourceread.ReadPodV1OrDie(podYaml)
+)
+
+type CopyToNodeRuntime struct {
+	PerNodePodRuntime *pernodepod.PerNodePodRuntime
+
+	FileSources []string
+}
+
+func (r *CopyToNodeRuntime) Run(ctx context.Context) error {
+	secret, secretDataKeyToFilename, err := r.secretFromFileSources()
+	if err != nil {
+		return fmt.Errorf("unable to create secret: %w", err)
+	}
+	secret.Name = "source-data"
+	secret.Namespace = r.PerNodePodRuntime.NamespacePrefix
+	r.PerNodePodRuntime.KubeClient.CoreV1().Secrets(secret.Namespace).Create(ctx, secret, metav1.CreateOptions{})
+
+	prePodHookFn := func(ctx context.Context, namespaceName string) (pernodepod.CleanUpFunc, error) {
+		secret.Name = "source-data"
+		secret.Namespace = namespaceName
+		if _, err := r.PerNodePodRuntime.KubeClient.CoreV1().Secrets(secret.Namespace).Create(ctx, secret, metav1.CreateOptions{}); err != nil {
+			return nil, fmt.Errorf("failed to create source-data: %w", err)
+		}
+		fmt.Fprintf(r.PerNodePodRuntime.Out, "Created --namespace=%v secrets/%v\n", namespaceName, secret.Name)
+
+		cleanupFn := func(ctx context.Context) {
+			if err := r.PerNodePodRuntime.KubeClient.CoreV1().Secrets(secret.Namespace).Delete(ctx, secret.Name, metav1.DeleteOptions{}); err != nil {
+				fmt.Fprintf(r.PerNodePodRuntime.ErrOut, "failed to cleanup source-data: %w", err)
+			}
+		}
+		return cleanupFn, nil
+	}
+
+	copyCommands := []string{}
+	for source, destination := range secretDataKeyToFilename {
+		sourcePath := filepath.Join("/source-data", source)
+		destPath := filepath.Join("/host-root", destination)
+		parentDir := filepath.Dir(destPath)
+		copyCommands = append(copyCommands, fmt.Sprintf("mkdir -p %q", parentDir))
+		copyCommands = append(copyCommands, fmt.Sprintf("cp --dereference -fr %q %q", sourcePath, destPath))
```

Comment:
Not fatal, but the problem with generating shell script here is that if any of the source/dest have shell metacharacters we won't be properly handling escaping.  There's e.g. https://pkg.go.dev/github.com/alessio/shellescape

We could also generate a pod per file to copy and just directly inject `cp` into the commands, although that's a bit messier with the semantic to make parent directories here.

Reply:
fixed